projectorganizer/src/prjorg-utils.c

```diff
@@ -241,3 +242,193 @@ GtkWidget *menu_item_new(const gchar *icon_name, const gchar *label)
+		doc_basename = g_path_get_basename(doc->file_name);
+
+		if (patterns_match(header_patterns, doc_basename))
+			is_header = TRUE;
+		else if (patterns_match(source_patterns, doc_basename))
+			is_header = FALSE;
+		else
+			known_type = FALSE;
+
+		g_free(doc_basename);
+	}
+
+	if (known_type)
+	{
+		GSList *elem = NULL, *list = NULL;
+		guint i = 0;
+
+		/* check open documents */
+		foreach_document(i)
+		{
+			gchar *filename;
+			filename = document_index(i)->file_name;
+			list = g_slist_prepend(list, filename);
+		}
+		found_name = try_find_header_source(doc->file_name, is_header, list, header_patterns, source_patterns);
+		g_slist_free(list);
+		list = NULL;
+
+		/* check document directory */
+		if (!found_name)
+		{
+			gchar *utf8_doc_dir;
+			gchar *locale_doc_dir;
+
+			utf8_doc_dir = g_path_get_dirname(doc->file_name);
+			locale_doc_dir = utils_get_locale_from_utf8(utf8_doc_dir);
+
+			list = utils_get_file_list(locale_doc_dir, NULL, NULL);
+			foreach_list (elem, list)
+			{
+				gchar *full_name;
+				full_name = g_build_filename(locale_doc_dir, elem->data, NULL);
+				SETPTR(full_name, utils_get_utf8_from_locale(full_name));
+				SETPTR(elem->data, full_name);
+			}
+			found_name = try_find_header_source(doc->file_name, is_header, list, header_patterns, source_patterns);
+			g_slist_foreach(list, (GFunc) g_free, NULL);
+			g_slist_free(list);
+			g_free(utf8_doc_dir);
+			g_free(locale_doc_dir);
+			list = NULL;
+		}
+
+		/* check project files */
+		if (!found_name && prj_org)
+		{
+			foreach_slist(elem, prj_org->roots)
+			{
+				GHashTableIter iter;
+				gpointer key, value;
+				PrjOrgRoot *root = elem->data;
+
+				list = NULL;
+				g_hash_table_iter_init(&iter, root->file_table);
+				while (g_hash_table_iter_next(&iter, &key, &value))
+					list = g_slist_prepend(list, key);
+				found_name = try_find_header_source(doc->file_name, is_header, list, header_patterns, source_patterns);
+				g_slist_free(list);
+				if (found_name)
+					break;
+			}
+		}
+	}
+
+	g_slist_foreach(header_patterns, (GFunc) g_pattern_spec_free, NULL);
+	g_slist_free(header_patterns);
+	g_slist_foreach(source_patterns, (GFunc) g_pattern_spec_free, NULL);
+	g_slist_free(source_patterns);
+
+	return found_name;
+}
+
+
+/* sets filetype of header to match source
+ * changes filetype only if document is a header */
+void set_header_filetype(GeanyDocument * doc)
+{
+	GSList * header_patterns;
+	if (!doc || !doc->file_name)
+		return;
+
+	if (prj_org)
+	{
+		header_patterns = get_precompiled_patterns(prj_org->header_patterns);
+	}
+	else
+	{
+		/* use default patterns when project isn't open */
+		gchar ** patterns = g_strsplit(PRJORG_PATTERNS_HEADER, " ", -1);
+		header_patterns = get_precompiled_patterns(patterns);
+		g_strfreev(patterns);
+	}
+
+	gchar * doc_basename = g_path_get_basename(doc->file_name);
+	gboolean is_header = patterns_match(header_patterns, doc_basename);
+	gchar * full_name = is_header ? find_header_source(doc) : NULL;
```

Comment:
Just a style thing - could you move the declarations of the 3 variables to the beginning of the function and only assign the variables here?

Apart from this change the patch looks good to me.

Reply:
I'll make the change.  Do you want the commits squashed again?